File: src/main/java/com/now/nowbot/model/JsonData/Covers.java
package com.now.nowbot.model.JsonData;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;

@JsonIgnoreProperties(ignoreUnknown = true)
public class Covers {
    @JsonProperty("cover")
    String cover;
    @JsonProperty("cover@2x")
    String cover2x;
    @JsonProperty("card")
    String card;
    @JsonProperty("card@2x")
    String card2x;
    @JsonProperty("list")
    String list;
    @JsonProperty("list@2x")
    String list2x;
    @JsonProperty("slimcover")
    String slimcover;
    @JsonProperty("slimcover@2x")
    String slimcover2x;

    public String getCover() {
        return cover;
    }

    public void setCover(String cover) {
        this.cover = cover;
    }

    public String getCover2x() {
        return cover2x;
    }

    public void setCover2x(String cover2x) {
        this.cover2x = cover2x;
    }

    public String getCard() {
        return card;
    }

    public void setCard(String card) {
        this.card = card;
    }

    public String getCard2x() {
        return card2x;
    }

    public void setCard2x(String card2x) {
        this.card2x = card2x;
    }

    public String getList() {
        return list;
    }

    public void setList(String list) {
        this.list = list;
    }

    public String getList2x() {
        return list2x;
    }

    public void setList2x(String list2x) {
        this.list2x = list2x;
    }

    public String getSlimcover() {
        return slimcover;
    }

    public void setSlimcover(String slimcover) {
        this.slimcover = slimcover;
    }

    public String getSlimcover2x() {
        return slimcover2x;
    }

    public void setSlimcover2x(String slimcover2x) {
        this.slimcover2x = slimcover2x;
    }

    @Override
    public String toString() {
        final StringBuilder sb = new StringBuilder("Covers{");
        sb.append("cover='").append(cover).append('\'');
        sb.append(", cover2x='").append(cover2x).append('\'');
        sb.append(", card='").append(card).append('\'');
        sb.append(", card2x='").append(card2x).append('\'');
        sb.append(", list='").append(list).append('\'');
        sb.append(", list2x='").append(list2x).append('\'');
        sb.append(", slimcover='").append(slimcover).append('\'');
        sb.append(", slimcover2x='").append(slimcover2x).append('\'');
        sb.append('}');
        return sb.toString();
    }
}

File: src/main/java/com/now/nowbot/model/osu/Covers.kt
package com.now.nowbot.model.osu

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonProperty

@JsonInclude(JsonInclude.Include.NON_NULL) @JsonIgnoreProperties(ignoreUnknown = true)
data class Covers (
    @field:JsonProperty("cover")
    val cover: String = "",

    @field:JsonProperty("cover@2x")
    val cover2x: String = "",

    @field:JsonProperty("card")
    val card: String = "",

    @field:JsonProperty("card@2x")
    val card2x: String = "",

    @field:JsonProperty("list")
    val list: String = "",

    @field:JsonProperty("list@2x")
    val list2x: String = "",

    @field:JsonProperty("slimcover")
    val slimcover: String = "",

    @field:JsonProperty("slimcover@2x")
    val slimcover2x: String = "",
) {
    
    // https://discord.com/channels/188630481301012481/1097318920991559880/1432996516934455378
    //
    //       ╭────() @YumeMuzi: .......
    //    ╭─────╮ peppy
    //    │ ppy │ and this will get you banned.
    //    ╰─────╯

    @get:JsonProperty("fullsize")
    val fullsize: String
        get() = list.replace("@2x", "").replace("list", "fullsize")

    override fun toString(): String {
        return "Covers(cover=$cover, cover2x=$cover2x, card=$card, card2x=$card2x, list=$list, list2x=$list2x, slimcover=$slimcover, slimcover2x=$slimcover2x)"
    }

    companion object {
        enum class CoverType {

            FULL_SIZE, CARD, CARD_2X, COVER, COVER_2X, LIST, LIST_2X, SLIM_COVER, SLIM_COVER_2X;

            companion object {
                fun getCovetType(string: String?): CoverType {
                    return when (string?.replace(" ", "")) {
                        "fullsize", "full", "raw", "r", "f", "background", "b" -> FULL_SIZE
                        "list", "l", "l1" -> LIST
                        "list2", "list2x", "list@2x", "l2" -> LIST_2X
                        "c", "card", "c1" -> CARD
                        "card2", "card2x", "card@2x", "c2" -> CARD_2X
                        "slim", "slimcover", "s", "s1" -> SLIM_COVER
                        "slim2", "slim2x", "slim@2x", "slimcover2", "slimcover2x", "slimcover@2x", "s2" -> SLIM_COVER_2X
                        "2", "cover2", "cover2x", "cover@2x", "o2" -> COVER_2X
                        null -> COVER
                        else -> COVER
                    }
                }

                fun Covers.getString(type: CoverType): String {
                    return when (type) {
                        LIST -> list
                        LIST_2X -> list2x
                        CARD -> card
                        CARD_2X -> card2x
                        SLIM_COVER -> slimcover
                        SLIM_COVER_2X -> slimcover2x
                        COVER_2X -> cover2x
                        COVER -> cover
                        FULL_SIZE -> list.replace("@2x", "").replace("list", "fullsize")
                    }
                }
            }
        }

        fun getCoverFromCacheID(beatmapsetID: Long, cacheID: Long? = null): Covers {

            val prefix = "https://assets.ppy.sh/beatmaps/${beatmapsetID}/covers/"

            val suffix = if (cacheID != null) {
                ".jpg?${cacheID}"
            } else {
                ".jpg"
            }

            return Covers(
                "${prefix}cover${suffix}",
                "${prefix}cover@2x${suffix}",
                "${prefix}card${suffix}",
                "${prefix}card@2x${suffix}",
                "${prefix}list${suffix}",
                "${prefix}list@2x${suffix}",
                "${prefix}silmcover${suffix}",
                "${prefix}silmcover@2x${suffix}",
            )
        }
    }
}
